stl/inc/xstring

```diff
@@ -2815,6 +2815,51 @@ private:
     }
 
 public:
+#ifdef __cpp_lib_containers_ranges
+    template <_Container_compatible_range<_Elem> _Rng>
+    constexpr basic_string(from_range_t, _Rng&& _Range, const _Alloc& _Al = _Alloc())
+        : _Mypair(_One_then_variadic_args_t{}, _Al) {
+        auto&& _Alproxy = _GET_PROXY_ALLOCATOR(_Alty, _Getal());
+        _Container_proxy_ptr<_Alty> _Proxy(_Alproxy, _Mypair._Myval2);
+        _Tidy_init();
+
+        if constexpr (_RANGES sized_range<_Rng>) {
+            const auto _Count = _Convert_size<size_type>(_To_unsigned_like(_RANGES size(_Range)));
+
+            if constexpr (contiguous_iterator<decltype(_RANGES _Ubegin(_Range))> && //
+                          same_as<_RANGES range_value_t<_Rng>, _Elem>) {
+                if (_Count != 0) {
+                    const auto _Ptr = _STD to_address(_RANGES _Ubegin(_Range));
+                    assign(_Ptr, _Count);
+                }

```

Comment:
This would be a useful optimization, but it's complicated by the fact that using the traits operations may be user-observable. In any case, I'd like to see such an optimization work for `basic_string<char>(const char8_t*, const char8_t*)` as well as for `basic_string<char>(from_range_t, basic_string<char8_t>&)`. Filing issue #2901 to investigate this as a followup.